### PR TITLE
Generate texture atlases at runtime

### PR DESCRIPTION
**Pull request by peppsac**
1st commit:
    Texture atlas usually perform better than individual textures.
    It also allows to reduce the number of Sampler2D to the number of layer,
    instead of the number of textures.
    
    I didn't merged all the layers in a single atlas, because updating this
    huge texture is slow (at least using the naive way of building a THREE.Texture
    from an atlas. It may be feasible to access directly WebGL functions like
    glTexSubImage2D).
    
    Currently the texture packer is naive and probably only works for our test
    data, but it should be easy to implement a more robust solution.
    
    Last but not least, without this commit iTowns crashes on my phone :)

~The 2nd commit leverages on the 1st and implements color textures transition.~ Edit: animation moved to a later PR

~Note that this PR is not quite ready for merging (layer deletion not implemented etc...) but you can test it and give feedback.~ Edit: ready!

Thanks!

### CODE REVIEW

**Comment by elemoine** [comment]
FWIW OpenLayers uses the bin packing algorithm described in http://www.blackpawn.com/texts/lightmaps/. See https://github.com/openlayers/openlayers/blob/master/src/ol/style/atlas.js for the OpenLayers implementation.

**Comment by peppsac** [comment]
@elemoine yep, this algorithm seems to be widly used. I was just too lazy to do it so I adapted a simple existing code. 
Note that we're not too constrained anyway: our textures are all power of 2 (mostly 256x256 pixels) but as we need to add a border around them to prevent color bleeding, we end up with 258x258 pixels.
And as the end result must be a power of 2 texture as well, we often waste a lot of space (e.g: 2 256x256 textures need to be stored in a 1024x512 atlas)

**Comment by qdnguyen** [comment]
>Texture atlas usually perform better than individual textures.
It also allows to reduce the number of Sampler2D to the number of layer,
instead of the number of textures.

Other solution is to use mega texture (virtual texture ) and update it through glTexSubImage2D, but i dont know how to support all LoDs of wmts with mega texture.

**Comment by peppsac** [comment]
@qdnguyen yeah, mega texture is an option, but I'm not familiar with them so I'll stick with atlases :)

**Comment by peppsac** [comment]
OK, this PR is ready now :)

**Comment by peppsac** [comment]
Updated and added a debug tool to see the atlas from the selected tile.

**Comment by gchoqueux** [comment]
little problem with OBB 
capture du 2017-06-09 15-40-51

**Comment by peppsac** [comment]
Updated: fixed the OBB bug and an issue with `img` and `THREE.Texture`

**Comment by gchoqueux** [comment]
I do a performance test :
`view.controls.setRange(50000,true)`
time on master : 1.5s and Atlas 2.5s

**Comment by peppsac** [comment]
> time on master : 1.5s and Atlas 2.5s

What are you measuring exactly? Here's what I do:
```
console.log('>>>'); globeView.controls.setRange(50000,true).then(()=>{console.log('<<<');})
```
My branch is a bit slower (4.5sec vs 4 sec) but it's expected since I do a bit more work than the default `three.js` implementation.

**Comment by peppsac** [comment]
Also in the general case the bottleneck will probably be the network accesses.

**Comment by peppsac** [comment]
Rebased and ready to review.

File: src/Core/Scheduler/Providers/Fetcher.js
Comment (autra) [inline src/Core/Scheduler/Providers/Fetcher.js]: Not requiring it here, but we should probably make this configurable in the future (as it depends on server config).

File: utils/debug/Debug.js
Comment (autra) [inline utils/debug/Debug.js]: + something like `height: 100%; overflow: auto` to make it scrollable if it overflows?

(I would prefer `width: 25%` but it depends what you want to do)

Comment (autra) [reply utils/debug/Debug.js]: Nice tool btw!

File: test/itowns-testing.js
Comment (autra) [inline test/itowns-testing.js:46]: I'm really starting to think we need something like [mock-browser](https://www.npmjs.com/package/mock-browser). It could be a good middleground between this and a headless browser wdyt? If you agree, I'll make a PR with this.

File: src/Renderer/Packer.js
Comment (autra) [inline src/Renderer/Packer.js:4]: Curious why you renamed `Math`?

File: src/Renderer/Packer.js
Comment (autra) [inline src/Renderer/Packer.js]: Could you document this method plz?

File: src/Renderer/LayeredMaterial.js
Comment (autra) [inline src/Renderer/LayeredMaterial.js]: Did I miss something or you are just doing a shallow copy of textures and offsetScale here?

Comment (autra) [inline src/Renderer/LayeredMaterial.js]: Did I miss something or you are just doing a shallow copy of textures and offsetScale here?

File: src/Renderer/LayeredMaterial.js
Comment (autra) [inline src/Renderer/LayeredMaterial.js]: I would say one of these 2 lines is redundant.

File: src/Renderer/LayeredMaterial.js
Comment (autra) [inline src/Renderer/LayeredMaterial.js:415]: What is this 0?

File: src/Renderer/LayeredMaterial.js
Comment (autra) [inline src/Renderer/LayeredMaterial.js]: nits `new Array(8)` ?

Comment (autra) [reply src/Renderer/LayeredMaterial.js]: Doesn't change anything here especially, but it's a matter of good habits for the most... dangerous part of js language :-) It works only because the writer of `Array` has coded for this use. In general, I'd like to see the `constructor starts with uppercase and should be called with new, normal function starts with lowercase and are called normally` rule enforced in itowns.

File: src/Renderer/LayeredMaterial.js
Comment (autra) [inline src/Renderer/LayeredMaterial.js]: Looks like you want to use a [default parameter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters) value for `updateFragmentShader` instead.

File: src/Core/Prefab/GlobeView.js
Comment (autra) [inline src/Core/Prefab/GlobeView.js]: That's cool!

File: src/Renderer/AtlasBuilder.js
Comment (autra) [inline src/Renderer/AtlasBuilder.js]: Can you document this function plz? (I struggle to understand it without it, and some comments below, maybe).

Comment (autra) [reply src/Renderer/AtlasBuilder.js]: My understanding is that you pack all the `images` into one big texture (with the help of the canvas thing below) and that uvs contains the individual coordinates. Is that right?

File: src/Renderer/AtlasBuilder.js
Comment (autra) [inline src/Renderer/AtlasBuilder.js]: Why do you need to sort? (I feel like I don't really understand your data layout)

File: src/Renderer/LayeredMaterial.js
Comment (autra) [inline src/Renderer/LayeredMaterial.js]: I guess we can remove these?

**Review by autra — CHANGES_REQUESTED** [review]

**Comment by peppsac** [comment]
Thanks for the review @autra
I won't reply to your comments here but instead I'll amend the code to fix all the issues you noted.

File: test/itowns-testing.js
Comment (peppsac) [reply test/itowns-testing.js:46]: Worth a try at least :)

File: src/Renderer/LayeredMaterial.js
Comment (peppsac) [reply src/Renderer/LayeredMaterial.js:415]: Because all coords of one atlas will have the same level/zoom anyway

File: examples/planar.html
Comment (autra) [inline examples/planar.html]: Do we need the diffs on this file in this PR?

File: src/Process/LayeredMaterialNodeProcessing.js
Comment (autra) [inline src/Process/LayeredMaterialNodeProcessing.js]: If i'm not mistaken, this function always returns an Array of size 1, or did I miss something?

Comment (autra) [reply src/Process/LayeredMaterialNodeProcessing.js]: It seems it is the case. Maybe we can then avoid the `for...off` below?

Comment (autra) [reply src/Process/LayeredMaterialNodeProcessing.js]: Ok I forgot about PM. I get it.

File: src/Process/LayeredMaterialNodeProcessing.js
Comment (autra) [inline src/Process/LayeredMaterialNodeProcessing.js]: Then `updateFragmentShader` should not be considered as private, and be renamed accordingly.

Comment (autra) [reply src/Process/LayeredMaterialNodeProcessing.js]: Can we indicate to initNodeImageryTexturesFromParent the correct sequence from the start? It would save the trouble of calling setSequence here, and would remove the need of the updateFragment flag in setSequence.

File: src/Renderer/AtlasBuilder.js
Comment (autra) [reply src/Renderer/AtlasBuilder.js]: I meant a block comment on top explaining the purpose of this method, how it achieves its goal and what it returns. I'm not even requiring you use jsdoc tags, as I know you like them so much (but do use `/**` to make it appear in the generated doc though plz).

File: src/Renderer/AtlasBuilder.js
Comment (autra) [inline src/Renderer/AtlasBuilder.js]: nits: one `image` is enough.

File: src/Renderer/LayeredMaterial.js
Comment (autra) [inline src/Renderer/LayeredMaterial.js]: I don't really see the point of this comment here? But anyway, maybe we want a map instead then.

File: src/Renderer/LayeredMaterial.js
Comment (autra) [inline src/Renderer/LayeredMaterial.js:355]: Why can't we create more than 14 samplers if `maxTexturesUnits` is bigger than 16?

File: src/Renderer/Packer.js
Comment (autra) [reply src/Renderer/Packer.js:4]: Ok Forgot about `Math` in js

File: src/Process/LayeredMaterialNodeProcessing.js
Comment (autra) [inline src/Process/LayeredMaterialNodeProcessing.js:27]: So this is true if c is strictly inside parentCoords, right? But with PM it's possible that c overlaps with one of the parent texture without being entirely in it right? We simply ignore this case?

File: src/Process/LayeredMaterialNodeProcessing.js
Comment (autra) [inline src/Process/LayeredMaterialNodeProcessing.js:76]: What was the problem?

File: src/Renderer/LayeredMaterial.js
Comment (autra) [reply src/Renderer/LayeredMaterial.js]: (Not requiring it here though)

**Review by autra — CHANGES_REQUESTED** [review]
A few comments, many questions :-) Overall this looks a lot simpler than previous code.

Thanks!

File: src/Renderer/LayeredMaterial.js
Comment (peppsac) [reply src/Renderer/LayeredMaterial.js]: Sorry I forgot to fix this one.
Is this rule available in eslint?

File: src/Process/LayeredMaterialNodeProcessing.js
Comment (peppsac) [reply src/Process/LayeredMaterialNodeProcessing.js:27]: (you're commenting on existing code that I simply updated)

And I *think* a PM texture will always be fully contained in its parent textures which are PM as well (would need to double-check this though)

File: src/Renderer/LayeredMaterial.js
Comment (autra) [reply src/Renderer/LayeredMaterial.js]: it's http://eslint.org/docs/rules/new-cap, currently disabled in our conf (but apparently they have exceptions for `Array`, so let's forget about this one)

Comment (autra) [reply src/Renderer/LayeredMaterial.js]: And it's currently disabled because we don't follow this rule at all :-/

File: src/Process/LayeredMaterialNodeProcessing.js
Comment (peppsac) [reply src/Process/LayeredMaterialNodeProcessing.js]: IMHO it's private to itowns, not to `LayeredMaterial`, hence the `_` prefix.

But I can remove it if you prefer.

File: examples/planar.html
Comment (peppsac) [reply examples/planar.html]: No sorry, I'll remove this.

File: src/Process/LayeredMaterialNodeProcessing.js
Comment (peppsac) [reply src/Process/LayeredMaterialNodeProcessing.js:76]: The code under the `if` assumes a xbil texture:
```
const { min, max } = OGCWebServiceHelper.ioDXBIL.computeMinMaxElevation(
```

File: src/Renderer/LayeredMaterial.js
Comment (peppsac) [reply src/Renderer/LayeredMaterial.js]: (again, this is existing code)

I'll remove it.

File: src/Renderer/LayeredMaterial.js
Comment (peppsac) [reply src/Renderer/LayeredMaterial.js:355]: (this code was there before as well)

I think the reasonning is that sometimes `Capabilities.getMaxTextureUnitsCount();` reported incorrect values, so the `16` was a reasonnable maximum.

This could be investigated and maybe removed...

File: src/Renderer/AtlasBuilder.js
Comment (peppsac) [reply src/Renderer/AtlasBuilder.js]: You're not requiring them, but our linter does.

File: src/Process/LayeredMaterialNodeProcessing.js
Comment (peppsac) [reply src/Process/LayeredMaterialNodeProcessing.js]: Thanks for the suggestion: I fixed the code using your idea.

File: src/Renderer/AtlasBuilder.js
Comment (autra) [inline src/Renderer/AtlasBuilder.js]: Thanks!

**Review by autra — DISMISSED** [review]

**Comment by autra** [comment]
LGTM, thanks!

**Comment by nosy-b** [comment]
Ok I check to see performance issues. It should be slower with texture atlas as it implies more JS slow canvas work. I tell you before 5:30 pm!

**Comment by nosy-b** [comment]
Ok, I didn't see a significant speed difference which is cool (although I didn't measure it precisely) but the memory footprint is much more, going above 1.2 Go after 45 sec of active navigation vs 0.5 Go with the master.
I would suggest to work on that after the release.

**Comment by peppsac** [comment]
@nosy-b :

> the memory footprint is much more, going above 1.2 Go after 45 sec of active navigation vs 0.5 Go with the master.

That's interesting. How did you measure this? (using dev tools memory tab or something else ?)

**Comment by nosy-b** [comment]
With the OS task manager (config win + chrome). It looks like creating texture combining others increases the memory footprint constantly, maybe you keep references to original tiles and new combining textures

**Comment by autra** [comment]
> With the OS task manager

Not sure it's a reliable way of measuring memory footprint. Navigators keeps some chunks of memory allocated from time to time for future use, or maybe it didn't GC immediately because your system still had some memory left etc. I suggest to use firefox  or chrome memory tab in devtools. In firefox, you can force a GC, CC and reclaim as much memory as it is possible in the `about:memory` page. I think it will yield more accurate results.

**Comment by peppsac** [comment]
So, using the task manager of Chrome (see https://developers.google.com/web/tools/chrome-devtools/memory-problems/) I could indeed confirm that the total memory used by my iTowns tab was growing.

I found and fixed 2 different bugs related to memory management, but the issue was still there.

And then I found these 2 bugs: https://bugs.chromium.org/p/chromium/issues/detail?id=242215 and https://bugs.chromium.org/p/chromium/issues/detail?id=626082 which seem related to our issue.

The second bug mentions that the issue has been (partially?) fixed 2 weeks ago :)

Testing with Chrome 61 confirmed that the problem seems fixed.

I'll try to see if I can mitigate the issue (I can maybe reused old canvas instead of creating new ones)...

**Comment by peppsac** [comment]
I just pushed 2 different approaches to limit the number of canvas used and thus the memleak.

**Comment by autra** [comment]
@peppsac which approach gave you the best results? Maybe we can add a // XXX or // FIXME comment explaining why this change, and linking to the bug reports ? I'll see if I reproduce in firefox.

**Comment by peppsac** [comment]
The second approach creates less `canvas` but is more fragile since it depends on 2 `three.js` behavior:
- a texture is only uploaded once (which is true if its .version field never change, which is true in our case)
- an uploaded texture has a `dispose` listener

I don't mind depending on three.js behavior, but we should probably introduce a comment/tag (`// three.js version: r0.86`?) everytime we do this to make sure this is well tested when we upgrade three.js.

**Comment by autra** [comment]
I've measuring memory on Firefox, and couldn't reproduce. While memory usage with this PR is consistently slightly higher than on master (around 10%), I saw no leaks.

How I tested:
- I loaded Itowns
- Waited for globe to be loaded, then took a memory snapshot
- zoomed on Paris, waited for globe to be loaded, then took another snapshot.

On master, I'm around 195-200MB. On this branch, around 215-220MB. @peppsac is this slight increase expected?

Anyhow, that confirms the chromium bug I guess. @peppsac I let you choose the approach you prefer to limit memory usage, but I don't think we should block this PR because of this bug. Maybe we can add some comments, maybe also a warning in the console with a link to the bug report for chromium developers?

File: src/Core/Prefab/GlobeView.js
Comment (peppsac) [reply src/Core/Prefab/GlobeView.js]: And broken with the last patch, because I recycle the canvas :)

Comment (autra) [reply src/Core/Prefab/GlobeView.js]: Damn'!

**Review by autra — APPROVED** [review]

**Review by gchoqueux — CHANGES_REQUESTED** [review]
Wait before merge, please
I would like reviewing this PR

**Comment by gchoqueux** [comment]
there's a problem with video memory
This PR uses more video memory that master branch

statut |  Gpu system bytes | Gpu dedicated bytes | Gpu committed bytes
------------ | ------------ | ------------- | -------------
 Master init |  21M  |  157M  | 177M
 Master range 500000 |  19M  | 175M  | 191M
Atlas init  |  26M |  195M | 219M
 Atlas range 500000 |  19M  |  260 M |  276M

May be textures atlas isn't completely filled.

Capture textures in video memory, There are empty areas
capture
capture

**Comment by peppsac** [comment]
> Capture textures in video memory, There are empty areas

Textures must use power of 2 sizes, so these empty areas are expected.

>  This PR uses more video memory that master branch

Video memory usage is not an end goal in itself.
That being said, yes this PR increases memory usage, but on the other hand it simplifies shader code and texture management.
It also opens the way to implement feature like animated transition.

We can refine `AtlasBuilder` to improve memory usage but I prefered to first land a naive version that performs OK and then iterate from there.

**Comment by gchoqueux** [comment]
Unfortunate this isn't a small increase in memory. 
Which can be fixed by filling in the empty slots in atlas

**Comment by peppsac** [comment]
> Unfortunate this isn't a small increase in memory.

So what?
It's a compromise: increased mem usage vs shader cost and other params (I forgot texture sampler count in my previous comment). 

> Which can be fixed by filling in the empty slots in atlas

I won't fill empty slots in atlas, since it's not possible.
But we can reduce GPU-object count by merging multiple atlases into one for instance.
We can also investigate the need for the color-bleed-preventing pixel for PM texture... hmm maybe I should stack them vertically since they are consecutives images... I'll give it a try.

**Comment by peppsac** [comment]
Pushed an update that should bring back memory usage to something closer to master.

Changes implemented:
- if a layer has a single image -> then skip the atlas mechanism and directly return a `THREE.Texture` as `TextureLoader` used to do for us
- if a layer has several textures: currently it only happens for PM layers so I added a preparation step to avoid the need to add a pixel line between individual images. So now a PM layer has this kind of atlas:
capture d ecran de 2017-07-19 15-19-44

@gchoqueux WDYT?

**Comment by gchoqueux** [comment]
I'll review, don't merge before my review, please

**Comment by gchoqueux** [comment]
I look at it this morning

**Comment by gchoqueux** [comment]
When you do serveral back and forth, there is a memory leak in the Vram. (Chrome windows v59)

`globeView.controls.setRange(500000,true)` and `globeView.controls.setRange(25100000,true) `

The dispose texture is more complicated?

**Comment by gchoqueux** [comment]
there's VRAM leak on  Chrome windows v61

**Comment by peppsac** [comment]
Updated to rebase on master and added a fix for `getXbilTextureByUrl`

Regarding the leak: IMHO as long as the textures are all released (= three.js texture count doesn't keep increasing) I don't see how the bug could be in itowns :-/

**Comment by peppsac** [comment]
(moved the `fix` commit to #401 )

**Comment by peppsac** [comment]
Branch updated.
I re-enabled a temporary debug tool to see which texture are used for each tile (only in index.html demo).

There's more work to do, but are we interested in getting this merged or not?
It should help with #654  and #583 but I'd like some feedback.

**Comment by autra** [comment]
I confirm that it helps a lot for #654. Planar examples works perfectly on a sony xperia tablet. Globe nearly works: I needed to use OPENSM.json layer: when near the ground, it works perfectly, when far away, texture are not displaying (but that might be because of the layer, I don't know).

Orthographic doesn't work anymore on both desktop and tablet with this PR though...

tested with both firefox and chrome

At least the Firefox doesn't crash any more.

**Comment by peppsac** [comment]
Self-closing this one. I'll reopen it later, when I have some time to work on it.

**Pull request closed without merging**
